src/Data/Text/Encoding.hs

```diff
@@ -187,31 +233,274 @@ decodeLatin1 bs = withBS bs $ \fp len -> runST $ do
+  else tds
+
+-- | Create a 'Text' value from the contents of a 'TextDataStack'.
+stackToText ::
+#if defined(ASSERTS)
+  HasCallStack =>
+#endif
+  TextDataStack -> Text
+stackToText (TextDataStack stack sLen)
+  | sLen > 0 = runST $
+    do
+      dst <- A.new sLen
+      let g (dat : dataStack') tLen' =
+              (case dat of
+                Left (Text arr0 off utf8Len) -> do
+                  let dstOff = tLen' - utf8Len
+                  A.copyI utf8Len dst dstOff arr0 off
+                  pure dstOff
+                Right bs@(B.length -> utf8Len) -> do
+                  let dstOff = tLen' - utf8Len
+                  withBS bs $ \ fp _ ->
+                    unsafeIOToST . unsafeWithForeignPtr fp $ \ src ->
+                      unsafeSTToIO $ A.copyFromPointer dst dstOff src utf8Len
+                  pure dstOff) >>= g dataStack'
+          g _ _ = pure ()
+      g stack sLen
+      arr <- A.unsafeFreeze dst
+      pure $ Text arr 0 sLen
+  | otherwise = empty
+
+-- | Decode a 'ByteString' in the context of what has been already been decoded.
+--
+-- The 'ByteString' is validated against the 'Utf8ValidState' using the rules
+-- governing 'validateNextUtf8Chunk'. The longest valid UTF-8 prefix is added
+-- to the input 'TextDataStack' which is returned with the end position of the
+-- valid prefix, and either the resulting 'Utf8ValidState'
+-- (@Right Utf8ValidState@) or the position of the of the first (potentially)
+-- valid byte after the invalid bytes with remainder of the 'ByteString'
+-- (@Left (Int, ByteString)@).
+decodeNextUtf8Chunk ::
+#if defined(ASSERTS)
+  HasCallStack =>
+#endif
+  ByteString
+  -> Utf8ValidState
+  -> TextDataStack
+  -> ((Int, Either (Int, ByteString) Utf8ValidState), TextDataStack)
+decodeNextUtf8Chunk bs s tds =
+  case validateNextUtf8Chunk bs s of
+    (len, res) ->
+      let stackedData'
+            | len >= 0 =
+              let stackedData@(TextDataStack stack' sLen') =
+                    foldl (\ tds'@(TextDataStack stack sLen) bs'@(B.length -> bLen) ->
+                      if bLen > 0
+                      then TextDataStack (Right bs' : stack) $ sLen + bLen
+                      else tds'
+                      ) tds $ partialUtf8CodePoint s
+              in
+              if len > 0
+              then TextDataStack (Right (B.take len bs) : stack') $ sLen' + len
+              else stackedData
+            | otherwise = tds
+      in
+      ( ( len
+        , case res of
+            Left pos -> Left (pos, B.drop pos bs)
+            Right s' -> Right s'
+        )
+      , stackedData'
+      )
+
+-- | Decode a 'ByteString' against a start 'Utf8ValidState' with an empty
+-- 'TextDataStack'.
+--
+-- @decodeUtf8Chunk bs = 'decodeNextUtf8Chunk' bs 'startUtf8ValidState' 'emptyStack'@
+decodeUtf8Chunk :: ByteString -> ((Int, Either (Int, ByteString) Utf8ValidState), TextDataStack)
+decodeUtf8Chunk bs = decodeNextUtf8Chunk bs startUtf8ValidState emptyStack
```

Comment:
A couple of reasons. I didn't want to deal with resizing the buffer, because the exact size of the buffer needed isn't known until the caller request the resulting text (by calling `stackToText`). There's also no guarantee that the caller will request the text, so it saves from having an unnecessary buffer created. Also, I think I ran into the "because type variable ... would escape its scope" error when I tried the buffer.